### PR TITLE
#436 | Resending messages from backup

### PR DESCRIPTION
**Pull request by tmmi**

### CODE REVIEW

File: hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java [hunk 1]
```diff
@@ -58,33 +66,83 @@ public BackupMessagesLoader(BrokerMessageProducer brokerMessageProducer,
...
+    }
+
+    private boolean sendMessageIfNeeded(Message message, Optional<Topic> topic, String contextName) {
+        if (topic.isPresent() && isNotStale(message)) {
+            sendMessage(message, topic.get());
+            return true;
+        } else {
+            logger.warn("Not {} stale message {} {} {}", contextName, message.getId(), topic.get().getQualifiedName(),
+                    new String(message.getData(), Charset.defaultCharset()));
+            return true;
```
Comment (druminski) [inline hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java]: we probably don't want return true here, also log message message is misleading and maybe we should distinct scenario with stale message from no topic info

Comment (dankraw) [inline hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java]: +1

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/config/Configs.java [hunk 2]
```diff
@@ -94,6 +94,7 @@
     MESSAGES_LOCAL_STORAGE_ENABLED("frontend.messages.local.storage.enabled", false),
     MESSAGES_LOCAL_STORAGE_DIRECTORY("frontend.messages.local.storage.directory", Files.createTempDir().getAbsolutePath()),
     MESSAGES_LOCAL_STORAGE_MAX_AGE_HOURS("frontend.messages.local.storage.max.age.hours", 72),
+    MESSAGES_LOCAL_STORAGE_MAX_RESENT("frontend.messages.local.storage.max.resent", 5),
```
Comment (dankraw) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/config/Configs.java]: maybe: `MESSAGES_LOCAL_STORAGE_MAX_RESEND_RETRIES`

File: hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java [hunk 3]
```diff
@@ -43,6 +47,10 @@
     private final Trackers trackers;
     private final int secondsToWaitForTopicsCache;
     private final int messageMaxAgeHours;
+    private final int resentSleep;
+    private final int maxResent;
+
+    private final AtomicReference<ConcurrentLinkedQueue<Pair<Message, Topic>>> toResent = new AtomicReference<>();
```
Comment (dankraw) [inline hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java]: toResent -> toResend

File: hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java [hunk 4]
```diff
@@ -58,33 +66,83 @@ public BackupMessagesLoader(BrokerMessageProducer brokerMessageProducer,
...
+                List<Pair<Message, Topic>> retryMessages = new ArrayList(toResent.getAndSet(new ConcurrentLinkedQueue<>()));
+                resendMessages(retryMessages, retry);
+            }
+            try {
+                Thread.sleep(resentSleep);
+            } catch (InterruptedException e) {
+                logger.warn("Sleep interrupted", e);
+            }
+            retry++;
+        } while (toResent.get().size() > 0 && retry <= maxResent);
```
Comment (dankraw) [inline hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java]: shouldn't this be `retry < maxResent`?

File: hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java [hunk 5]
```diff
@@ -58,33 +66,83 @@ public BackupMessagesLoader(BrokerMessageProducer brokerMessageProducer,
         this.trackers = trackers;
         this.secondsToWaitForTopicsCache = config.getIntProperty(MESSAGES_LOADING_WAIT_FOR_TOPICS_CACHE);
         this.messageMaxAgeHours = config.getIntProperty(MESSAGES_LOCAL_STORAGE_MAX_AGE_HOURS);
+        this.resentSleep = config.getIntProperty(KAFKA_PRODUCER_ACK_TIMEOUT) + secondsToWaitForTopicsCache * 1000;
+        this.maxResent = config.getIntProperty(MESSAGES_LOCAL_STORAGE_MAX_RESENT);
```
Comment (dankraw) [inline hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java]: `maxResendRetries` maybe

File: hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java (hunk 4 above)
Comment (tmmi) [inline hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java]: <= seems ok; retry=0 is not resend

File: hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java (hunk 1 above)
Comment (tmmi) [inline hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java]: fixed

**Comment by wojtkiewicz** [comment]
:+1:

**Pull request closed without merging**